### PR TITLE
Pubsub ordering keys

### PR DESCRIPTION
**Pull request by kamalaboulhosn**
Copy ordering keys code to master branch. The performance regression for the non-ordering case has been fixed.

### CODE REVIEW

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 1]
```diff
@@ -85,15 +88,17 @@
   private final String topicName;
 
   private final BatchingSettings batchingSettings;
+  private final boolean enableMessageOrdering;
 
   private final Lock messagesBatchLock;
-  private MessagesBatch messagesBatch;
+  final Map<String, MessagesBatch> messagesBatches;
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: This variable should be private.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 2]
```diff
@@ -85,15 +88,17 @@
...
 
   private final AtomicBoolean activeAlarm;
 
   private final PublisherStub publisherStub;
 
   private final ScheduledExecutorService executor;
+  final SequentialExecutorService.CallbackExecutor sequentialExecutor;
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: This variable should be privte

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 3]
```diff
@@ -194,38 +210,72 @@ public String getTopicNameString() {
...
     messagesBatchLock.lock();
     try {
+      MessagesBatch messagesBatch = messagesBatches.get(orderingKey);
+      if (messagesBatch == null) {
+        messagesBatch = new MessagesBatch(batchingSettings, orderingKey);
+        messagesBatches.put(orderingKey, messagesBatch);
+      }
+
       batchesToSend = messagesBatch.add(outstandingPublish);
-      // Setup the next duration based delivery alarm if there are messages batched.
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: This comment should be restored by `setupAlarm()`

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 4]
```diff
@@ -194,38 +210,72 @@ public String getTopicNameString() {
...
     } finally {
       messagesBatchLock.unlock();
     }
 
     messagesWaiter.incrementPendingMessages(1);
 
-    if (!batchesToSend.isEmpty()) {
+    // For messages without ordering keys, it is okay to send batches without holding
+    // messagesBatchLock.
+    if (!batchesToSend.isEmpty() && orderingKey.isEmpty()) {
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: If ordering keys is turned on, should `publishAllWithoutInflight()` still happen?

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 5]
```diff
@@ -257,16 +307,53 @@ public void run() {
...
     try {
-      if (messagesBatch.isEmpty()) {
-        return;
+      for (MessagesBatch batch : messagesBatches.values()) {
+        if (!batch.isEmpty()) {
+          // TODO(kimkyung-goog): Do not release `messagesBatchLock` when publishing a batch. If
+          // it's released, the order of publishing cannot be guaranteed if `publish()` is called
+          // while this function is running. This locking mechanism needs to be improved if it
+          // causes any performance degradation.
+          publishOutstandingBatch(batch.popOutstandingBatch());
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Should this happen for non ordering keys scenarios?

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java (hunk 4 above)
Comment (kamalaboulhosn) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Looks like we shouldn't call it, no.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java (hunk 5 above)
Comment (kamalaboulhosn) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: I moved it out from under the lock. We should do the work, but it doesn't need to be done while the lock is held.

**Comment by kamalaboulhosn** [comment]
@sduskis I have removed the test that was problematic. It was the one that had been using the now-private elements of the class and so I think it is too dependent on the internal details to be able to consistently test something meaningful. Do you want to complete this review or do you want me to hand it off to someone else? Thanks!

**Comment by sduskis** [comment]
@kamalaboulhosn, please find someone else.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 6]
```diff
@@ -256,17 +308,62 @@ public void run() {
...
    */
   public void publishAllOutstanding() {
+    OutstandingBatch unorderedOutstandingBatch = null;
+    messagesBatchLock.lock();
+    try {
+      for (MessagesBatch batch : messagesBatches.values()) {
+        if (!batch.isEmpty()) {
+          if (!batch.orderingKey.isEmpty()) {
+            // For messages with an ordering key, we need to publish with messagesBatchLock held in
+            // order
```
Comment (manuelmenzella-google) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Nit: I think this break is unintentional.

**Pull request merged**